File: WordPuzzle.java
import student.*;
import java.util.ArrayList;
import java.util.Random;
// -------------------------------------------------------------------------
/**
 *  This program creates a WordPuzzle class which will manipulate string and arraylists of integers.
 * The purpose of this project is to learn how to create classes  and manipulate strings in Java.
 * In this project , I created WordPuzzle class , which methods can change message into 
 * puzzle and puzzle into message. First I create object using WordClass and then I access its 
 * method which can return puzzle (ArrayList of nested pairs of number) or message. The inputs in this object 
 * can be a message and a puzzle or one of each of them.
 * 
 *  @Adam Lech adaml8
 *  @2012.05.01
 */
public class WordPuzzle
{
    //~ Instance/static variables .............................................
	 // Instantiating variables which are the inputs of the class
      private String message;  
      private ArrayList<ArrayList<ArrayList<Integer>>> puzzle; 

    //~ Constructor ...........................................................
       
     // Gives a set of instructions to initialize object 
       
    // ----------------------------------------------------------
    /**
     * Creates a new WordPuzzle object.
     */
	//Empty Constructor
    public WordPuzzle() 
    {
        //we don't need to do anything for an empty constructor.
    }
    //This constructor initialize object if  input is message consisting of string characters
	// It uses setFromMessage() method to initialize object 
    public WordPuzzle(String message) 
                                      
    {
       
        
        this.setFromMessage(message);
        
    }
    //This constructor initialize object if  input is puzzle  consisting of ArrayList
	// It uses setFromPuzzle() method to initialize object 
    public WordPuzzle(ArrayList<ArrayList<ArrayList<Integer>>> puzzle) 
                                                                       
    {
       
         
         this.setFromPuzzle(puzzle); 
        
    }
    // This constructor initialize object if input is a messsage and puzzle
    public WordPuzzle(String message, ArrayList<ArrayList<ArrayList<Integer>>> puzzle) 
    {   
        this.message = message; 
        this.puzzle = puzzle ; 
        
    }
     // This constructor initialize object if input is a puzzle and message
    public WordPuzzle(ArrayList<ArrayList<ArrayList<Integer>>> puzzle, String message) 
    {  
        this.puzzle = puzzle; 
        this.message = message; 
        
        
    }  
    //getter method  for message 
    public String getMessage() 
    {
        return this.message; 
    }
    //sets the message
    public void setMessage(String message) 
    {
          this.setFromMessage(message); 
    }
    //getter method for puzzle
    public ArrayList<ArrayList<ArrayList<Integer>>> getPuzzle() 
    {
       return this.puzzle; 
    }
    //sets the puzzle 
    public void setPuzzle(ArrayList<ArrayList<ArrayList<Integer>>> puzzle) 
    {
        this.setFromPuzzle(puzzle) ; 
    }
   //sets the message and puzzle if the input is a message
    public void setFromMessage(String message) 
    {
        this.message = message; 
        puzzle = this.puzzleFromMessage(message); //uses puzzleFromMessage() method to set the puzzle
        
    }
     //sets the message and puzzle if the input is a puzzle 
    public void setFromPuzzle(ArrayList<ArrayList<ArrayList<Integer>>> puzzle)
    { 
        this.puzzle = puzzle; 
        message = this.messageFromPuzzle(puzzle); //uses messageFromPuzzle() to set message
        
        
    }
    //This method obtains ASCII code for  a given string character 
    public static int ord(char chr) 
    {
        return (int)chr;  
    }
    //defines a method which will transform a puzzle consisting of an ArrayList
	// of integers into a message 
    public String messageFromPuzzle(ArrayList<ArrayList<ArrayList<Integer>>> puzzle) 
                                                                                     
     {
        
            
            String message = "";  //creates an empty string to store a message
           

		   //loops over every ArrayList of an ArrayList which represents individual word in the message
            for(Integer i = 0 ; i <puzzle.size(); ++i) { 
               
                 ArrayList<ArrayList<Integer>> wordPuzzle = puzzle.get(i); //Creates an ArrayList which consists of ArrayList of pair numbers 
                                                                           // representing a word 
                 String word = "";  // creates an empty string to store a word
                 
                 
                 
                 //Starts nested loop which will go over every pair of Integers in an ArrayList representing a character 
                 for(Integer p = 0; p < wordPuzzle.size(); ++p) {
                        
                        ArrayList<Integer>  pair = wordPuzzle.get(p);  //Creates an ArrayList which consists of pair of integers sum of which represents a string character
                                                                       
                        int letterNumber = pair.get(0) + pair.get(1); //Adds pairs of integers which make up a string character
                        
                        char letter = (char) letterNumber; //Changes a letterNumber into  a string character
                        word= word + letter; //Adds string character into a word of message
                        
                }
           
                
                message =   message + " "  + word ; //adds word into a message
               
            }   
              String finalMessage = message.trim(); //As the result of looping there are some redundant space characters in the message 
                                                    // I used .trim() method to remove them. I also created a new string to store the final message
        return  finalMessage;  //returns message
        
        
    }

    public ArrayList<ArrayList<ArrayList<Integer>>> puzzleFromMessage(String message) //creates a method which will transform string into an an ArrayList     of pairs of integers representing this message
                                                                                   
     {
        
        /*# Do any work to initialize your class here. *///this create a random number generator, if you do 
        //r.nextInt(someNumber)
        //then it will give you an integers from 0-someNumber
        Random r = new Random();

        
        String[] words = message.split(" "); // this will be an array of Strings with each element being one of the
                                             // words from your message
       
         
        
       ArrayList<ArrayList<ArrayList<Integer>>> puzzle  =  new  ArrayList<ArrayList<ArrayList<Integer>>>(); //create a new ArrayList<ArrayList<ArrayList<Integer>>> to hold your final result
                                                                                                            
      
        
       
        
        //loops over all the words in words 
          for (Integer i =0;i <  words.length; ++i) {  
              
              String word = words[i]; //Creates a new string word which will store an individual  word from the words array
              ArrayList<ArrayList<Integer>> wordPairs = new ArrayList<ArrayList<Integer>>();  //creates an ArrayList<ArrayList<Integer>> wordPairs to hold the pairs  that belong to the current word in the message
                                                                                             
              //starts nested loop 
              for(Integer l = 0; l <word.length(); ++l) { 
                 
                 ArrayList<Integer> letterNumber = new  ArrayList<Integer>(); //Creates an ArrayList of integers which will store a pair of integers 
                                                                             // representing each character in the message
                 
                 Integer number = (int)word.charAt(l);  //obtains ASCII code number for each of the string characters in the message
                 Integer firstNumber = r.nextInt(number); //obtains a first number of a pair which is a random number 
                 Integer secondNumber = number -firstNumber; //obtains a second number of a pair which is a difference between number and first number
                 letterNumber.add(firstNumber) ; //adds first number to the pair of integers representing each string character in the message
                 letterNumber.add(secondNumber) ; //adds second number to the pair of integers representing each string character in the message 
                 wordPairs.add(letterNumber); //adds a pair of integers(letterNumber) to the wordPairs ArrayList which represents an individual word in the message
                 
                 
               }
               puzzle.add(wordPairs); // adds an ArrayList representing an individual word in the message to the final puzzle 
              
               
          
    }
      
      
      return puzzle; // returns puzzle 

    

}

    public static char chr(int chrVal) //This methods changes ASCII code number into a string character 
    {
        char r = (char) chrVal;
        return r;
    }
    
    public String toString() //toString() method which didn't work in my case because instructor had an older version of java than I had (1.7.0_03) 
    {
        String prettyPrint = "";
        String topLine;
        String bottomLine;
        String underLine;
        for (ArrayList<ArrayList<Integer>> wordArray : this.puzzle)
        {
            topLine = "";
            bottomLine = "";
            underLine = "";
            for (ArrayList<Integer> pair : wordArray)
            {
                topLine += padLeft(pair.get(0).toString(), 4) + "  ";
                bottomLine += "+" + padLeft(pair.get(1).toString(), 3) + "  ";
                underLine += "----  ";
            }
            prettyPrint += topLine + "\n" + bottomLine + "\n" + underLine + "\n\n\n";
        }
        return prettyPrint;
    }
    
    public static String padLeft(String s, int n) 
    {
        return String.format("%1$#" + n + "s", s);   
    }

 
    public static void main(String[] args) //main method which tests all the methods in this class; 
	
	
	// { System.out.println(WordPuzzle.getClass()): 
	
	
	
	
	
	
	
    {
        ArrayList<ArrayList<ArrayList<Integer>>> puzzle = new ArrayList<ArrayList<ArrayList<Integer>>>();
        ArrayList<ArrayList<Integer>>word = new ArrayList<ArrayList<Integer>>();
        ArrayList<Integer>pair = new ArrayList<Integer>();
        pair.add(61);
        pair.add(11);
        word.add(pair);
        pair = new ArrayList<Integer>();
        pair.add(93);
        pair.add(4);
        word.add(pair);
        pair = new ArrayList<Integer>();
        pair.add(73);
        pair.add(39);
        word.add(pair);
        pair = new ArrayList<Integer>();
        pair.add(78);
        pair.add(34);
        word.add(pair);
        pair = new ArrayList<Integer>();
        pair.add(17);
        pair.add(104);
        word.add(pair);
        puzzle.add(word);
        word = new ArrayList<ArrayList<Integer>>();
        pair = new ArrayList<Integer>();
        pair.add(23);
        pair.add(43);
        word.add(pair);
        pair = new ArrayList<Integer>();
        pair.add(11);
        pair.add(93);
        word.add(pair);
        pair = new ArrayList<Integer>();
        pair.add(65);
        pair.add(52);
        word.add(pair);
        pair = new ArrayList<Integer>();
        pair.add(20);
        pair.add(96);
        word.add(pair);
        pair = new ArrayList<Integer>();
        pair.add(66);
        pair.add(31);
        word.add(pair);
        pair = new ArrayList<Integer>();
        pair.add(86);
        pair.add(24);
        word.add(pair);
        pair = new ArrayList<Integer>();
        pair.add(40);
        pair.add(61);
        word.add(pair);
        pair = new ArrayList<Integer>();
        pair.add(102);
        pair.add(13);
        word.add(pair);
        pair = new ArrayList<Integer>();
        pair.add(50);
        pair.add(51);
        word.add(pair);
        puzzle.add(word);
        word = new ArrayList<ArrayList<Integer>>();
        pair = new ArrayList<Integer>();
        pair.add(73);
        pair.add(43);
        word.add(pair);
        pair = new ArrayList<Integer>();
        pair.add(28);
        pair.add(73);
        word.add(pair);
        pair = new ArrayList<Integer>();
        pair.add(8);
        pair.add(89);
        word.add(pair);
        pair = new ArrayList<Integer>();
        pair.add(31);
        pair.add(68);
        word.add(pair);
        pair = new ArrayList<Integer>();
        pair.add(77);
        pair.add(27);
        word.add(pair);
        pair = new ArrayList<Integer>();
        pair.add(24);
        pair.add(77);
        word.add(pair);
        pair = new ArrayList<Integer>();
        pair.add(42);
        pair.add(72);
        word.add(pair);
        pair = new ArrayList<Integer>();
        pair.add(15);
        pair.add(24);
        word.add(pair);
        pair = new ArrayList<Integer>();
        pair.add(64);
        pair.add(51);
        word.add(pair);
        puzzle.add(word);
        word = new ArrayList<ArrayList<Integer>>();
        pair = new ArrayList<Integer>();
        pair.add(25);
        pair.add(75);
        word.add(pair);
        pair = new ArrayList<Integer>();
        pair.add(7);
        pair.add(90);
        word.add(pair);
        pair = new ArrayList<Integer>();
        pair.add(10);
        pair.add(111);
        word.add(pair);
        pair = new ArrayList<Integer>();
        pair.add(17);
        pair.add(16);
        word.add(pair);
        puzzle.add(word);
        WordPuzzle wp = new WordPuzzle();
        
        WordPuzzle wp2 = new WordPuzzle(puzzle);
        System.out.println(wp2); 
        System.out.println(wp.messageFromPuzzle(puzzle));

    }
} 

File: WordPuzzle.cpp
#include <iostream> 
using std::cout; 
using std::cin; 
using std::endl; 

#include <vector> 
using std:: vector;

#include "WordPuzzle.h" //includes definition of class

#include <stdLib.h> 
#include <time.h> 
#include <sstream> 
using std::ostringstream; 
 
 WordPuzzle:: WordPuzzle() { 
    //Empty constructor
  }
  
 WordPuzzle:: WordPuzzle(string message) { 
 
   WordPuzzle::message = message; 
   puzzle = puzzleFromMessage(message); 
 }
WordPuzzle:: WordPuzzle(vector<vector<vector<int>>> puzzle) { 
     
	WordPuzzle::puzzle = puzzle; 
    message =  messageFromPuzzle(puzzle); 
} 
WordPuzzle:: WordPuzzle(vector<vector<vector<int>>> puzzle , string message) { 
    WordPuzzle::message = message; 
    WordPuzzle::puzzle = puzzle; 
} 

WordPuzzle:: WordPuzzle(string message , vector<vector<vector<int>>> puzzle) { 
     WordPuzzle::message = message; 
	 WordPuzzle::puzzle = puzzle; 
} 


  
string WordPuzzle::getMessage() { 

  return message; 
} 

vector<vector<vector<int>>> WordPuzzle:: getPuzzle() { 
  
  return puzzle; 
} 
string WordPuzzle::messageFromPuzzle(vector<vector<vector<int>>> puzzle) { 
       string message = ""; 
	   for (int i=0 ; i < puzzle.size(); i++) {  
	      
	      vector<vector<int>> wordPuzzle = puzzle.at(i); 
    	  string word = ""; 
	       
		  for(int j=0; j < wordPuzzle.size(); j++) { 
		     vector<int> pair = wordPuzzle.at(j); 
			 int letterNumber = pair.at(0) + pair.at(1); 
			 char letter = (char)letterNumber; 
			 
			 word.append(1,letter); 
			
		  }
		  message = message + " " + word; 
	   } 
	   string finalMessage =  trim(message); 
	   return finalMessage; 
} 
vector<vector<vector<int>>> WordPuzzle::puzzleFromMessage(string message) {
      srand(time(NULL)); 
	  vector<string> words = split(message, ' '); 
	  
	  vector<vector<vector<int>>> puzzle(0); 
	  for(int i=0; i <words.size(); i++) { 
	     
		 string word = words.at(i); 
		 vector<vector<int>> wordPairs(0); 
		 
		 for(int j=0; j<word.size(); j++) { 
		  
		  vector<int> letterPair(2); 
		  int number = (int) word.at(i); 
		  
		  int firstNumber = rand()%number; 
		  int secondNumber = number -firstNumber; 
		  
		  letterPair.insert(letterPair.begin() , firstNumber); 
		  letterPair.insert(letterPair.begin() + 1, secondNumber); 
		  wordPairs.insert(wordPairs.begin() + j , letterPair); 
		  
		 
		 } 
		 
		 puzzle.insert(puzzle.begin() +i , wordPairs); 
		 
	  }
	  
	  return puzzle; 
}
void WordPuzzle::setFromMessage(string message) { 
   
   WordPuzzle::message = message; 
   puzzle = puzzleFromMessage(message); 

} 

void WordPuzzle::setFromPuzzle(vector<vector<vector<int>>> puzzle) { 

    WordPuzzle::puzzle = puzzle; 
	message = messageFromPuzzle(puzzle); 
   
} 

void WordPuzzle::setMessage(string message) { 
   
     WordPuzzle::message = message; 
	 
} 
void WordPuzzle::setPuzzle(vector<vector<vector<int>>> puzzle) { 
     WordPuzzle::puzzle = puzzle; 
} 

string WordPuzzle::asString() { 
     
	 string prettyPrint = ""; 
	 string topLine; 
	 string bottomLine;
	 string underLine;
	 
	 for each ( vector<vector<int>> wordArray in puzzle  ) { 
	 
	      topLine = ""; 
		  bottomLine = ""; 
		  underLine = ""; 
		  
		  for each(vector<int> pair in wordArray) { 
		      string firstN = integerToString<int>(pair.at(0)); 
			  string secondN = integerToString<int>(pair.at(1)); 
		      topLine += padLeft(firstN , 4) + "  " ; 
			  bottomLine += "+" + padLeft(secondN , 3) + "  "; 
			  underLine += "----  "; 
		    } 
		prettyPrint += topLine  +"\n" + bottomLine + "\n" + underLine + "\n\n\n"; 
	    
	 } 
	 return prettyPrint; 
} 

string WordPuzzle::padLeft(string s , int n) { 
    
     
	 string ret = s; 
     if(s.size() < n) { 
		 
         ret.insert(0 ,n - s.size(), ' '); 
		
	  } 
	 else  { 
	   ret = s; 
     }
     return ret; 
	
	
 } 
//private methods
vector<string> WordPuzzle::split(string s, char c) { 
   int counter = 0; 
   string p = s; 
   int pos = p.find_first_of(c); 
   vector<string> retVector(0); 
   while(pos != -1) { 
   cout<< "pos is: " <<pos<< endl; 
        p = s.substr(0, pos); 
		
		s = s.substr(pos +1  , string::npos); 
		cout<< "s is: " << s << endl; 
		pos = s.find_first_of(c); 
		
		retVector.insert(retVector.begin() + ++counter, p); 
		if(pos == -1) {
		  retVector.insert(retVector.begin() + counter, s); 
		} 
     } 
   
   return retVector; 
}  

string WordPuzzle::trim(string s) { 
    if (s.size() == 0) { 
	   return s; 
	} 
    while(s.at(s.size() -1) == ' ') { 
        s.erase(s.end() -1); 
     } 
	 while(s.at(0) == ' ') { 
        s.erase(s.begin()); 
     } 
     return s; 
} 

template <class T> 
string WordPuzzle::integerToString(T n) { 
     ostringstream ss; 
	 ss << n; 
	 return ss.str(); 
} 

 